File: apps/translate-tencent_HY-MT/iOS/OfflineTranslator/Views/ResultView.swift
import SwiftUI

/// Result state (reference screenshot #3): source on top, streamed translation on the
/// bottom, with an action footer. Tapping the source returns to editing.
struct ResultView: View {
    @ObservedObject var vm: TranslationViewModel
    let onEditSource: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            // Source (top half)
            VStack(alignment: .leading, spacing: 12) {
                Button(action: onEditSource) {
                    ScrollView {
                        Text(vm.sourceText)
                            .font(.system(size: 20, weight: .regular))
                            .foregroundStyle(Theme.textPrimary)
                            .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)

                iconButton("speaker.wave.2") { vm.speak(vm.sourceText, language: vm.source) }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
            .padding(.bottom, 14)

            Rectangle()
                .fill(Theme.separator)
                .frame(height: 1)

            // Translation (bottom half)
            VStack(alignment: .leading, spacing: 12) {
                ScrollView {
                    translationText
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)

                footer
            }
            .padding(.top, 16)
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        }
    }

    @ViewBuilder
    private var translationText: some View {
        if vm.translatedText.isEmpty && vm.isTranslating {
            HStack(spacing: 6) {
                ProgressView().tint(Theme.textSecondary)
                Text("Translating on device…")
                    .font(.system(size: 15))
                    .foregroundStyle(Theme.textSecondary)
            }
        } else {
            (Text(vm.translatedText)
                + Text(vm.isTranslating ? " ▍" : "").foregroundColor(Theme.accent))
                .font(.system(size: 20, weight: .regular))
                .foregroundStyle(Theme.textPrimary)
        }
    }

    private var footer: some View {
        HStack(spacing: 20) {
            iconButton("speaker.wave.2") { vm.speak(vm.translatedText, language: vm.target) }

            Text("Alternatives")
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(Theme.textSecondary)
                .padding(.horizontal, 13)
                .padding(.vertical, 7)
                .background(Theme.surfaceRaised, in: Capsule())

            Spacer()

            iconButton("bookmark") {}                       // placeholder
            ShareLink(item: vm.translatedText) {
                Image(systemName: "square.and.arrow.up")
                    .font(.system(size: 18, weight: .regular))
                    .foregroundStyle(Theme.textPrimary)
            }
            .disabled(vm.translatedText.isEmpty)
            iconButton("doc.on.doc") { vm.copyTranslation() }
        }
    }

    private func iconButton(_ name: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Image(systemName: name)
                .font(.system(size: 18, weight: .regular))
                .foregroundStyle(Theme.textPrimary)
        }
        .buttonStyle(.plain)
    }
}

File: apps/translate-tencent_HY-MT/Android/app/src/main/java/ai/zetic/demo/offlinetranslator/ui/ResultView.kt
package ai.zetic.demo.offlinetranslator.ui

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.background
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.VolumeUp
import androidx.compose.material.icons.filled.BookmarkBorder
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.Share
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import ai.zetic.demo.offlinetranslator.theme.Theme

/**
 * Result state: source on top, streamed translation on the bottom, with an action footer.
 * Tapping the source returns to editing. Mirrors iOS `ResultView`.
 */
@Composable
fun ResultView(
    sourceText: String,
    translatedText: String,
    isTranslating: Boolean,
    onEditSource: () -> Unit,
    onSpeakSource: () -> Unit,
    onSpeakTranslation: () -> Unit,
    onShare: () -> Unit,
    onCopy: () -> Unit,
) {
    Column(Modifier.fillMaxSize()) {
        // Source (top half)
        Column(
            modifier = Modifier.fillMaxWidth().weight(1f).padding(bottom = 14.dp),
        ) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
                    .verticalScroll(rememberScrollState())
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        onClick = onEditSource,
                    ),
            ) {
                Text(text = sourceText, color = Theme.textPrimary, fontSize = 20.sp)
            }
            Spacer(Modifier.height(12.dp))
            IconButton(Icons.AutoMirrored.Filled.VolumeUp, onSpeakSource)
        }

        Box(Modifier.fillMaxWidth().height(1.dp).background(Theme.separator))

        // Translation (bottom half)
        Column(
            modifier = Modifier.fillMaxWidth().weight(1f).padding(top = 16.dp),
        ) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
                    .verticalScroll(rememberScrollState()),
            ) {
                TranslationText(translatedText, isTranslating)
            }
            Spacer(Modifier.height(12.dp))
            Footer(
                onSpeak = onSpeakTranslation,
                onShare = onShare,
                onCopy = onCopy,
                shareEnabled = translatedText.isNotEmpty(),
            )
        }
    }
}

@Composable
private fun TranslationText(translatedText: String, isTranslating: Boolean) {
    if (translatedText.isEmpty() && isTranslating) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            CircularProgressIndicator(
                color = Theme.textSecondary,
                strokeWidth = 2.dp,
                modifier = Modifier.size(16.dp),
            )
            Spacer(Modifier.width(8.dp))
            Text("Translating on device…", color = Theme.textSecondary, fontSize = 15.sp)
        }
    } else {
        val caretAlpha = if (isTranslating) blinkingAlpha() else 0f
        val text: AnnotatedString = buildAnnotatedString {
            append(translatedText)
            if (isTranslating) {
                withStyle(SpanStyle(color = Theme.accent.copy(alpha = caretAlpha))) {
                    append(" ▍")
                }
            }
        }
        Text(text = text, color = Theme.textPrimary, fontSize = 20.sp)
    }
}

@Composable
private fun blinkingAlpha(): Float {
    val transition = rememberInfiniteTransition(label = "caret")
    val alpha by transition.animateFloat(
        initialValue = 1f,
        targetValue = 0f,
        animationSpec = infiniteRepeatable(tween(600), RepeatMode.Reverse),
        label = "caretAlpha",
    )
    return alpha
}

@Composable
private fun Footer(
    onSpeak: () -> Unit,
    onShare: () -> Unit,
    onCopy: () -> Unit,
    shareEnabled: Boolean,
) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        IconButton(Icons.AutoMirrored.Filled.VolumeUp, onSpeak)
        Spacer(Modifier.width(20.dp))
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(percent = 50))
                .background(Theme.surfaceRaised)
                .padding(horizontal = 13.dp, vertical = 7.dp),
        ) {
            Text("Alternatives", color = Theme.textSecondary, fontSize = 13.sp, fontWeight = FontWeight.Medium)
        }
        Spacer(Modifier.weight(1f))
        IconButton(Icons.Filled.BookmarkBorder, onClick = { /* placeholder, inert like iOS */ })
        Spacer(Modifier.width(20.dp))
        IconButton(Icons.Filled.Share, onShare, enabled = shareEnabled)
        Spacer(Modifier.width(20.dp))
        IconButton(Icons.Filled.ContentCopy, onCopy)
    }
}

@Composable
private fun IconButton(icon: ImageVector, onClick: () -> Unit, enabled: Boolean = true) {
    val tint = if (enabled) Theme.textPrimary else Theme.textTertiary
    Box(
        modifier = Modifier
            .size(28.dp)
            .clickable(
                enabled = enabled,
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
                onClick = onClick,
            ),
        contentAlignment = Alignment.Center,
    ) {
        Icon(imageVector = icon, contentDescription = null, tint = tint, modifier = Modifier.size(18.dp))
    }
}
